### PR TITLE
Minio helm chart improvements for user and policy creation

### PR DESCRIPTION
**Pull request by jescalona-lf**
## Description
Added support to helm chart to define policies in addition to default ones and create user with a secretKey based on existing secret.

## Motivation and Context
Looking forward to automate fine-grained permissions for users in Minio's helm chart and additionally, wanted to created new users based on existing K8S secrets.

## How to test this PR?

1. Clone repo and checkout branch
2. Change directory to cloned repo
3. Create 2 secrets in K8S:
```
$ kubectl create secret bucket1-secret --from-literal=secretKey=bucket1
$ kubectl create secret bucket2-secret --from-literal=secretKey=bucket2
```
4. Deploy Minio cluster using the following test-values.yml file:
```
$ cat test-values.yml
buckets:
  - name: bucket1
    policy: none
    purge: false
    versioning: false
  - name: bucket2
    policy: none
    purge: false
    versioning: false


policies:
  - name: bucket1
    statements:
      - resources: 
          - 'arn:aws:s3:::bucket1/*'
          - 'arn:aws:s3:::bucket1'
        actions:
          - "s3:ListBucket"
          - "s3:ListBucketMultipartUploads"          
          - "s3:AbortMultipartUpload"
          - "s3:GetObject"
          - "s3:DeleteObject"
          - "s3:PutObject"
          - "s3:ListMultipartUploadParts"
          - "s3:CreateBucket"
          - "s3:DeleteBucket"
          - "s3:GetBucketLocation"
  - name: bucket2
    statements:
      - resources: 
          - "arn:aws:s3:::bucket2/*"
          - "arn:aws:s3:::bucket2"
        actions:          
          - "s3:GetObject"
          - "s3:GetBucketLocation"
          - "s3:ListBucket"
          - "s3:ListBucketMultipartUploads"          

users:
  - accessKey: bucket1AccessKey
    existingSecretKey: secretKey
    existingSecret: bucket1-secret
    policy: bucket1
  - accessKey: bucket2AccessKey
    existingSecretKey: secretKey
    existingSecret: bucket2-secret
    policy: bucket2
$ helm install minio-test . -f values.yaml -f test-values.yml
```
5. Validate policies were created. 
6. Verify that users bucket1AccessKey and bucket2AccessKey
7. Verify that users can authenticate with the secrets created previously

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [x] Documentation updated
- [ ] Unit tests added/updated

### CODE REVIEW

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
<html><body><h1>503 Service Unavailable</h1>
No server is available to handle this request.
</body></html>

Deleting image locally

File: helm/minio/templates/_helper_create_policy.txt [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+# connectToMinio
+# Use a check-sleep-check loop to wait for MinIO service to be available
+connectToMinio() {
+  SCHEME=$1
+  ATTEMPTS=0 ; LIMIT=29 ; # Allow 30 attempts
+  set -e ; # fail if we can't read the keys.
+  ACCESS=$(cat /config/rootUser) ; SECRET=$(cat /config/rootPassword) ;
+  set +e ; # The connections to minio are allowed to fail.
+  echo "Connecting to MinIO server: $SCHEME://$MINIO_ENDPOINT:$MINIO_PORT" ;
+  MC_COMMAND="${MC} config host add myminio $SCHEME://$MINIO_ENDPOINT:$MINIO_PORT $ACCESS $SECRET" ;
```
Comment (harshavardhana) [inline helm/minio/templates/_helper_create_policy.txt]:
```suggestion
  MC_COMMAND="${MC} alias set myminio $SCHEME://$MINIO_ENDPOINT:$MINIO_PORT $ACCESS $SECRET" ;
```

File: helm/minio/templates/_helper_policy.tpl [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+  "Version": "2012-10-17",
+  "Statement": [
+{{- range $i, $statement := .statements }}
+    {
+      "Effect": "Allow",
+      "Principal": {
+        "AWS": [
+          "*"
+        ]
+      },
```
Comment (harshavardhana) [inline helm/minio/templates/_helper_policy.tpl]:
```suggestion
      "Principal": {
        "AWS": [
          "*"
        ]
      },
```

IAM policies do not carry principals, remove please.

File: helm/minio/templates/configmap.yaml [hunk 3]
```diff
@@ -13,3 +13,9 @@ data:
 {{ include (print $.Template.BasePath "/_helper_create_bucket.txt") . | indent 4 }}
   add-user: |-
 {{ include (print $.Template.BasePath "/_helper_create_user.txt") . | indent 4 }}
+  add-policy: |-
+{{ include (print $.Template.BasePath "/_helper_create_policy.txt") . | indent 4 }}
+{{- range .Values.policies }}
+  {{ .name }}.json: |-
+{{ include (print $.Template.BasePath "/_helper_policy.tpl") . | indent 4 }}
+{{ end }}
```
Comment (harshavardhana) [inline helm/minio/templates/configmap.yaml]: newline missing.

File: helm/minio/templates/post-install-create-policy-job.yaml [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+                name: {{ template "minio.fullname" . }}
+            - secret:
+                name: {{ template "minio.secretName" . }}
+        {{- if .Values.tls.enabled }}
+        - name: cert-secret-volume-mc
+          secret:
+            secretName: {{ .Values.tls.certSecret }}
+            items:
+            - key: {{ .Values.tls.publicCrt }}
+              path: CAs/public.crt
```
Comment (harshavardhana) [inline helm/minio/templates/post-install-create-policy-job.yaml:66]: should this not be `/.mc/certs/CAs/public.crt` ?

File: helm/minio/templates/post-install-create-policy-job.yaml [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        volumeMounts:
+          - name: minio-configuration
+            mountPath: /config
+          {{- if .Values.tls.enabled }}
+          - name: cert-secret-volume-mc
+            mountPath: {{ .Values.configPathmc }}certs
+          {{ end }}
+        resources:
+{{ toYaml .Values.makePolicyJob.resources | indent 10 }}
+{{- end }}
```
Comment (harshavardhana) [inline helm/minio/templates/post-install-create-policy-job.yaml]: newline

File: helm/minio/values.yaml [hunk 6]
```diff
@@ -245,6 +245,59 @@ resources:
...
+#     - resources: 
+#         - 'arn:aws:s3:::example2'
+#         - 'arn:aws:s3:::example3'
+#       actions:
+#         - "s3:AbortMultipartUpload"
+#         - "s3:GetObject"
+#         - "s3:DeleteObject"
+#         - "s3:PutObject"
+#         - "s3:ListMultipartUploadParts"
+#     - resources:
```
Comment (harshavardhana) [inline helm/minio/values.yaml]: some of these resources need to have wildcard at the end to work can you update?

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Review by harshavardhana — APPROVED** [review]

**Review by harshavardhana — APPROVED** [review]

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Review by harshavardhana — APPROVED** [review]
Thanks for all the helpful changes.

**Comment by jescalona-lf** [comment]
It's my pleasure @harshavardhana

**Pull request merged**